src/components/Table.js

```diff
@@ -177,81 +197,233 @@ export default class Table extends React.Component<TableDefaultProps, TableProps
+      document.removeEventListener('keydown', this.keyDown);
+      document.removeEventListener('keyup', this.keyUp);
+    }
+  }
 
-      if (!selectedRowIds.includes(rowId)) {
-        // if the row is not already in the selection, add it
-        selectedRowIds.push(rowId);
-        this.props.onSelect(selectedRowIds, rowId);
-      } else if (selectedRowIds.length === 1 && this.props.noEmptySelection) {
-        // if this is the only remaining selection and we don't allow empty selection, just don't update
-        return false;
-      } else {
-        // if it is already in the selection, remove it
-        selectedRowIds.splice(selectedRowIds.indexOf(rowId), 1);
-        rowId = selectedRowIds[selectedRowIds.length - 1];
-        this.props.onSelect(selectedRowIds, rowId);
-      }
-      return true;
+  getFormattedSelection = (selection): string => {
+    // const { selection } = this.props;
+    if (Array.isArray(selection)) {
+      return selection[0];
+    } else if (typeof selection === 'string') {
+      return selection;
     }
+    return `${selection}`;
+  }
 
-    let selectedRowIds;
-    if (Array.isArray(this.props.selection)) {
-      selectedRowIds = this.props.selection[0];
-    } else if (typeof this.props.selection === 'string') {
-      selectedRowIds = this.props.selection;
-    } else {
-      selectedRowIds = '';
+  getSelectedIds = (newSelectedRowIndices: Set<number>): Array => {
+    const { sortedRows } = this.state;
+    const selectedRowArray = [];
+    newSelectedRowIndices.forEach((rowIndex) => {
+      selectedRowArray.push(rowIndex);
+    });
+    const selectedRowIds = selectedRowArray.map((rowIndex) => {
+      return sortedRows[rowIndex] ? `${sortedRows[rowIndex].id}` : null;
+    });
+    return selectedRowIds;
+  }
+
+  keyDown = (e: KeyboardEvent) => {
+    if (e.shiftKey || e.ctrlKey || e.metaKey) {
+      const shiftKeyDown = e.shiftKey;
+      const ctrlKeyDown = e.ctrlKey || e.metaKey;
+      this.setState({ shiftKeyDown, ctrlKeyDown });
+    }
+  }
+
+  keyUp = (e: KeyboardEvent) => {
+    const shiftKeyUp = e.key === 'Shift';
+    const ctrlKeyUp = e.key === 'Control' || e.key === 'Meta';
+    if (shiftKeyUp || ctrlKeyUp) {
+      this.setState({ shiftKeyDown: false, ctrlKeyDown: false });
+    }
+  }
+
+  rowSelect = (rowData: any): boolean => {
+    const {
+      anchorRowIndex,
+      ctrlKeyDown,
+      selectedRowIndices: prevSelectedRowIndices,
+      shiftKeyDown,
+      sortedRows,
+    } = this.state;
+    const shiftKeyPressed = shiftKeyDown;
+    const ctrlKeyPressed = ctrlKeyDown;
+    const newSelectedRowIndices = prevSelectedRowIndices;
+
+    const { multiSelect, noEmptySelection, onSelect, activeRowId } = this.props;
+
+    if (!rowData) {
+      return false;
     }
 
-    // if this row is the currently selected row,
-    if (selectedRowIds === rowId) {
+    if (multiSelect) {
+      if (shiftKeyPressed) {
+        if (anchorRowIndex === null) {
+          this.setState({
```

Comment:
By putting it up here, it's serving as a safety check as well. The rest of the shift stuff will hit an error if `anchorRowIndex` is null and the behavior is different in the ctrl scenario, all though one could argue. This would only come into place for tables that allow no selections. In that case, one could argue that clicking control should also behave the same way - If there are no selections and you select the table using control, that row should become selected. I focused on shift because it seemed it would be clearly frustrating if users tried to click a range and nothing happened and an anchor row is required for that, but I can't think of a good reason not to support the same with control. I would just move that line above the `shiftKeyPressed` so it can still serve as a safety net.